Bump docker-py to 4.0.2, and run tests from upstream repository

thaJeztah opened the pull request:
closes https://github.com/moby/moby/pull/37939
addresses https://github.com/moby/moby/issues/36415
closes https://github.com/moby/moby/pull/39239

This removes all the installation steps for docker-py from the
Dockerfile, and instead builds the upstream Dockerfile, and runs
docker-py tests in a container.

To test;

```
make test-docker-py

...

Removing bundles/

---> Making bundle: dynbinary (in bundles/dynbinary)
Building: bundles/dynbinary-daemon/dockerd-dev
Created binary: bundles/dynbinary-daemon/dockerd-dev

---> Making bundle: test-docker-py (in bundles/test-docker-py)
---> Making bundle: .integration-daemon-start (in bundles/test-docker-py)
Using test binary docker
Starting dockerd
INFO: Waiting for daemon to start...
.
INFO: Building docker-sdk-python3:3.7.0...
sha256:686428ae28479e9b5c8fdad1cadc9b7a39b462e66bd13a7e35bd79c6a152a402
INFO: Starting docker-py tests...
============================= test session starts ==============================
platform linux -- Python 3.6.8, pytest-4.1.0, py-1.8.0, pluggy-0.9.0
rootdir: /src, inifile: pytest.ini
plugins: timeout-1.3.3, cov-2.6.1
collected 359 items

tests/integration/api_build_test.py .......s....
....
```

relates to https://github.com/docker/docker-py/issues/2310

thaJeztah commented on Dockerfile:
```diff
@@ -199,25 +187,12 @@ RUN apt-get update && apt-get install -y \
 	g++-mingw-w64-x86-64 \
 	net-tools \
 	pigz \
-	python-backports.ssl-match-hostname \
```
Need to check if there's other dependencies that were only there for docker-py

thaJeztah replied:
@chris-crone any other reps that stand out and might be docker-py specific?

thaJeztah commented on hack/ci/janky:
```diff
@@ -12,7 +12,4 @@ bash <(curl -s https://codecov.io/bash) \
 hack/make.sh \
 	binary-daemon \
 	dynbinary \
-	test-integration-flaky \
```
temporarily removed these so that we only run the docker-py tests here

thaJeztah commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,29 @@ set -e
+
 (
 	bundle .integration-daemon-start
+	if [ -z "${DOCKER_SOCK}" ]; then
```
perhaps instead, we can check if `DOCKER_HOST` starts with `unix://` and strip that

thaJeztah replied:
Did even better; test can now run against remote daemons as well (which could be localhost); I changed the script to start the container with `--network=host` if `DOCKER_HOST` uses `tcp://` 🎉

thaJeztah commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,29 @@ set -e
 
-# subshell so that we can export PATH without breaking other things
+# The commit or tag to use for testing
+# TODO docker 17.06 cli client used in CI fails to build using a sha;
```
Looks like a bug; it fails when using a commit, but passes if I use a tag 🤔

thaJeztah commented:
Yay, it works 🎉 

```
17:37:53 ======== 348 passed, 5 skipped, 4 xfailed, 2 xpassed in 384.42 seconds ========
```

thaJeztah committed:
Makefile: Allow passing DOCKER_TEST_HOST and TESTDEBUG to container

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on Makefile:
```diff
@@ -50,9 +50,11 @@ DOCKER_ENVS := \
 	-e DOCKER_USERLANDPROXY \
 	-e DOCKERD_ARGS \
 	-e TEST_INTEGRATION_DIR \
+	-e TESTDEBUG \
```
Added these, so that they can be passed to the container 🎉

thaJeztah commented:
Removed "WIP" - I think this is ready for prime-time 🎉 

ping @tianon @chris-crone @cpuguy83 PTAL 🤗

thaJeztah commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,47 @@ set -e
 
+		tcp)
+		    # run container in host-mode networking so that it can connect to the
+		    # daemon from the current networking namespace (e.g., to connect to localhost)
```
Ah, booh, some spaces sneaked their way in; I'll update

tianon commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,47 @@ set -e
-		dockerPy="$DEST/docker-py"
-		git clone https://github.com/docker/docker-py.git "$dockerPy"
-	}
+	docker_host_scheme=`echo "${DOCKER_HOST}" | cut -d: -f1 -`
```
```console
$ DOCKER_HOST='tcp://foo:bar/baz'
$ echo "${DOCKER_HOST%%:*}"
tcp
```

```suggestion
	docker_host_scheme="${DOCKER_HOST%%:*}"
```

tianon commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,47 @@ set -e
+			run_opts="-v ${DOCKER_HOST#unix://}:/var/run/docker.sock"
+			;;
 
+		tcp)
```
You could also skip the `docker_host_scheme` variable entirely and just use globs in these case blocks, ala:

```bash
case "$DOCKER_HOST" in
    unix://*) ...
    tcp://*) ...
...
```

thaJeztah replied:
Oh, nice! Yes definitely gonna do that. My stackoverflow skills failed to find a good example for this (went searching for "find/strip scheme/prefix" 🤷‍♂️

tianon commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,47 @@ set -e
-	test_env PYTHONPATH="$dockerPy" py.test --junitxml="$DEST/results.xml" "$dockerPy/tests/integration"
+	case "${docker_host_scheme}" in
+		unix)
+			# trim the tcp:// scheme, and bind-mount the docker socket into the container
```
I think you meant "trim the unix:// scheme" here, not "tcp://".

thaJeztah replied:
😙🤫

tianon commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,47 @@ set -e
+		tcp)
+			# run container in host-mode networking so that it can connect to the
+			# daemon from the current networking namespace (e.g., to connect to localhost)
+			run_opts=--network=host -e DOCKER_HOST=${DOCKER_HOST}
```
Doesn't this need quotes to prevent Bash from trying to run `-e` as a command?

thaJeztah replied:
Good catch; I think I moved the `-e` here after initial testing with a remote daemon 😊

thaJeztah commented:
Looks like some issues with an image we're using in CI:

```
19:13:43 FAIL: docker_cli_pull_test.go:273: DockerSuite.TestPullWindowsImageFailsOnLinux
19:13:43 
19:13:43 assertion failed: expected error to contain "cannot be used on this platform", got "\nCommand:  /usr/local/cli/docker pull microsoft/nanoserver\nExitCode: 1\nError:    exit status 1\nStdout:   Using default tag: latest\n\nStderr:   Error response from daemon: manifest for microsoft/nanoserver:latest not found: manifest unknown: manifest unknown\n\n\nFailures:\nExitCode was 1 expected 0\nExpected no error"
19:13:45 
```

thaJeztah commented:
@tianon thanks! Updated; ptal 🤗

tianon reviewed (approved):
LGTM :+1:

thaJeztah commented:
@cpuguy83 LGTY?

chris-crone reviewed (approved):
LGTM

@shin- put up a PR to merge the [Dockerfiles for docker-py](https://github.com/docker/docker-py/pull/2222) which will mean that this requires changes. It's not merged yet though.

thaJeztah commented:
Ah, I see, so it would have to be 

```
-f tests/Dockerfile
```

instead of

```
-f Dockerfile-py3
```

thaJeztah commented:
rebased to trigger CI

thaJeztah commented:
```
01:49:56   Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f9b083c40b8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
01:49:56   Could not find a version that satisfies the requirement appdirs==1.4.3 (from -r requirements.txt (line 1)) (from versions: )
01:49:56 No matching distribution found for appdirs==1.4.3 (from -r requirements.txt (line 1))

thaJeztah commented:
Hmm, looks like it's failing to install the dependencies;

```
20:21:35 Step 5/9 : RUN pip install -r requirements.txt
20:21:35  ---> Running in 7a87168413c5
20:21:35 Ignoring cryptography: markers 'python_version == "3.3"' don't match your environment
20:21:35 Ignoring pypiwin32: markers 'sys_platform == "win32" and python_version < "3.6"' don't match your environment
20:21:35 Ignoring pypiwin32: markers 'sys_platform == "win32" and python_version >= "3.6"' don't match your environment
20:21:35 Ignoring urllib3: markers 'python_version == "3.3"' don't match your environment
20:21:35 Collecting appdirs==1.4.3 (from -r requirements.txt (line 1))
20:21:35   WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7fa810864588>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
20:21:35   WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7fa810864128>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
20:21:35   WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7fa810864860>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
20:21:35   WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7fa810864048>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
20:21:35   WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7fa810864dd8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)': /simple/appdirs/
20:21:35   ERROR: Could not find a version that satisfies the requirement appdirs==1.4.3 (from -r requirements.txt (line 1)) (from versions: none)
20:21:35 ERROR: No matching distribution found for appdirs==1.4.3 (from -r requirements.txt (line 1))
20:21:35 The command '/bin/sh -c pip install -r requirements.txt' returned a non-zero code: 1
```

Let me kick if again to see if it's just a temporary glitch

coolljt0725 commented:
ping @thaJeztah  janky failed :)

thaJeztah committed:
Bump docker-py to 4.0.2, and run tests from upstream repository

This removes all the installation steps for docker-py from the
Dockerfile, and instead builds the upstream Dockerfile, and runs
docker-py tests in a container.

To test;

```
make test-docker-py

...

Removing bundles/

---> Making bundle: dynbinary (in bundles/dynbinary)
Building: bundles/dynbinary-daemon/dockerd-dev
Created binary: bundles/dynbinary-daemon/dockerd-dev

---> Making bundle: test-docker-py (in bundles/test-docker-py)
---> Making bundle: .integration-daemon-start (in bundles/test-docker-py)
Using test binary docker
Starting dockerd
INFO: Waiting for daemon to start...
.
INFO: Building docker-sdk-python3:3.7.0...
sha256:686428ae28479e9b5c8fdad1cadc9b7a39b462e66bd13a7e35bd79c6a152a402
INFO: Starting docker-py tests...
============================= test session starts ==============================
platform linux -- Python 3.6.8, pytest-4.1.0, py-1.8.0, pluggy-0.9.0
rootdir: /src, inifile: pytest.ini
plugins: timeout-1.3.3, cov-2.6.1
collected 359 items

tests/integration/api_build_test.py .......s....
....
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
rebased on top of https://github.com/moby/moby/pull/39239 to get the latest docker-py version

thaJeztah commented:
Pushed a temporary commit to only run the docker-py tests on janky

thaJeztah commented:
Getting close; this looks like something that needs an update in that test, and either skip the test or ignore that error;

```
15:22:54 =================================== FAILURES ===================================
15:22:54 ___________________ SwarmTest.test_init_swarm_data_path_addr ___________________
15:22:54 docker/api/client.py:261: in _raise_for_status
15:22:54     response.raise_for_status()
15:22:54 /usr/local/lib/python3.6/site-packages/requests/models.py:940: in raise_for_status
15:22:54     raise HTTPError(http_error_msg, response=self)
15:22:54 E   requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http+docker://localhost/v1.35/swarm/init
15:22:54 
15:22:54 During handling of the above exception, another exception occurred:
15:22:54 tests/integration/api_swarm_test.py:254: in test_init_swarm_data_path_addr
15:22:54     assert self.init_swarm(data_path_addr='eth0')
15:22:54 tests/integration/base.py:127: in init_swarm
15:22:54     return self._init_swarm(self.client, **kwargs)
15:22:54 tests/integration/base.py:94: in _init_swarm
15:22:54     '127.0.0.1', listen_addr=helpers.swarm_listen_addr(), **kwargs
15:22:54 docker/utils/decorators.py:34: in wrapper
15:22:54     return f(self, *args, **kwargs)
15:22:54 docker/api/swarm.py:170: in init_swarm
15:22:54     return self._result(response, json=True)
15:22:54 docker/api/client.py:267: in _result
15:22:54     self._raise_for_status(response)
15:22:54 docker/api/client.py:263: in _raise_for_status
15:22:54     raise create_api_error_from_http_exception(e)
15:22:54 docker/errors.py:31: in create_api_error_from_http_exception
15:22:54     raise cls(e, response=response, explanation=explanation)
15:22:54 E   docker.errors.APIError: 400 Client Error: Bad Request ("interface eth0 has more than one IPv6 address (2001:db8:1::242:ac11:2 and fe80::42:acff:fe11:2)")
```

thaJeztah committed:
docker-py: skip flaky tests

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on hack/make/test-docker-py:
```diff
@@ -3,18 +3,51 @@ set -e
+
+# custom options to pass py.test
+# TODO remove this skip once https://github.com/docker/docker-py/pull/2380 is merged
+: ${PY_TEST_OPTIONS:="-k 'not test_init_swarm_data_path_addr'"}
```
See if this works; manually skipping `test_init_swarm_data_path_addr`

thaJeztah replied:
Hm.. nope; looks like it's now selecting on filename, not test name 🤔 

```
12:31:31 ========================= no tests ran in 0.01 seconds =========================

12:31:31 ERROR: file not found: test_init_swarm_data_path_addr'
```

thaJeztah replied:
Looks like `--deselect` is what I was looking for

thaJeztah committed:
docker-py: don't build --quiet is TESTDEBUG is set

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Whoop; it's green; 

```
19:05:11 = 369 passed, 10 skipped, 1 deselected, 3 xfailed, 2 xpassed in 406.70 seconds =
```

Removing the last commit to run all tests again

thaJeztah commented:
More flakiness? https://jenkins.dockerproject.org/job/Docker-PRs/54837/console

```
10:33:31 FAIL: docker_cli_swarm_unix_test.go:58: DockerSwarmSuite.TestSwarmNetworkPluginV2
10:33:31 
10:33:31 Creating a new daemon
10:33:31 [d2a8590156c0a] waiting for daemon to start
10:33:31 [d2a8590156c0a] waiting for daemon to start
10:33:31 [d2a8590156c0a] daemon started
10:33:31 
10:33:31 Creating a new daemon
10:33:31 [d3b4f063eba44] waiting for daemon to start
10:33:31 [d3b4f063eba44] waiting for daemon to start
10:33:31 [d3b4f063eba44] daemon started
10:33:31 
10:33:31 waited for 13.17522997s (out of 30s)
10:33:31 docker_cli_swarm_unix_test.go:91:
10:33:31     // wait to ensure all containers have exited before removing the plugin. Else there's a
10:33:31     // possibility of container exits erroring out due to plugins being unavailable.
10:33:31     waitAndAssert(c, defaultReconciliationTimeout, reducedCheck(sumAsIntegers, d1.CheckActiveContainerCount, d2.CheckActiveContainerCount), checker.Equals, 0)
10:33:31 docker_utils_test.go:435:
10:33:31     c.Assert(v, checker, args...)
10:33:31 ... obtained int = 1
10:33:31 ... expected int = 0
10:33:31 ... output: "54354283d360\n"
10:33:31 
10:33:31 waited for 30.111316047s (out of 30s)
10:33:31 [d2a8590156c0a] Stopping daemon
10:33:31 [d2a8590156c0a] exiting daemon
10:33:31 [d2a8590156c0a] Daemon stopped
10:33:31 [d3b4f063eba44] Stopping daemon
10:33:31 [d3b4f063eba44] exiting daemon
10:33:31 [d3b4f063eba44] Daemon stopped
```

thaJeztah commented:
Janky failing on another flaky; `TestAPISwarmNodeDrainPause` https://github.com/moby/moby/issues/23516

```
14:37:41 FAIL: docker_api_swarm_node_test.go:75: DockerSwarmSuite.TestAPISwarmNodeDrainPause
14:37:41 
14:37:41 Creating a new daemon
14:37:41 [d0bcbf80b8df2] waiting for daemon to start
14:37:41 [d0bcbf80b8df2] waiting for daemon to start
14:37:41 [d0bcbf80b8df2] daemon started
14:37:41 
14:37:41 Creating a new daemon
14:37:41 [d0dd452ee477a] waiting for daemon to start
14:37:41 [d0dd452ee477a] waiting for daemon to start
14:37:41 [d0dd452ee477a] daemon started
14:37:41 
14:37:41 waited for 961.666608ms (out of 30s)
14:37:41 waited for 9.089831ms (out of 30s)
14:37:41 waited for 17.422688ms (out of 30s)
14:37:41 waited for 1.789291659s (out of 30s)
14:37:41 waited for 8.880152ms (out of 30s)
14:37:41 docker_api_swarm_node_test.go:103:
14:37:41     waitAndAssert(c, defaultReconciliationTimeout*2, reducedCheck(sumAsIntegers, d1.CheckActiveContainerCount, d2.CheckActiveContainerCount), checker.Equals, instances)
14:37:41 docker_utils_test.go:435:
14:37:41     c.Assert(v, checker, args...)
14:37:41 ... obtained int = 2
14:37:41 ... expected int = 1
14:37:41 ... output: "067710d1439b\n7e874f57ab0a\n"
14:37:41 
14:37:41 waited for 1m0.118110186s (out of 1m0s)
14:37:41 [d0bcbf80b8df2] Stopping daemon
14:37:41 [d0bcbf80b8df2] exiting daemon
14:37:41 [d0bcbf80b8df2] Daemon stopped
14:37:41 [d0dd452ee477a] Stopping daemon
14:37:41 [d0dd452ee477a] exiting daemon
14:37:41 [d0dd452ee477a] Daemon stopped
14:38:16 
```

thaJeztah commented:
Opened some PRs upstream that might help getting this working;

- https://github.com/docker/docker-py/pull/2382 Adjust --platform tests for changes in docker engine
- https://github.com/docker/docker-py/pull/2380 xfail test_init_swarm_data_path_addr
- https://github.com/docker/docker-py/pull/2379 Update to python 3.7 (buster) and use build-args

thaJeztah commented:
Temporarily bumped to docker-py master, to see if the switch to Debian Buster makes it build again. It's not tagged yet, and there's still some other fixes not yet merged upstream

thaJeztah commented:
This looks like it may be a regression

```
23:35:35 === RUN   TestHealthKillContainer
23:35:47 --- FAIL: TestHealthKillContainer (11.49s)
23:35:47     health_test.go:57: assertion failed: error is not nil: Error response from daemon: Invalid signal: SIGUSR1

thaJeztah committed:
docker-py: use host-network for nested build of docker-py

When building this image docker-in-docker, the DNS in the environment
may not be usable for the build-container, causing resolution to fail:

```
02:35:31 W: Failed to fetch http://deb.debian.org/debian/dists/jessie/Release.gpg  Temporary failure resolving 'deb.debian.org'
```

This patch detects if we're building from within a container, and if
so, skips creating a networking namespace for the build by using
`--network=host`.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
OK, I think this should be ready to go now! (finally 😓)

```
11:53:09 = 367 passed, 10 skipped, 3 deselected, 4 xfailed, 1 xpassed in 391.32 seconds =
```

@tianon @cpuguy83 @vdemeester PTAL

thaJeztah commented:
Windows RS1 is failing to wipe the workspace https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/25913/console

```
12:36:11 Building remotely on azure-windows-rs1-3 (windows-rs1) in workspace /gopath/src/github.com/docker/docker
12:36:11 No credentials specified
12:36:11 Wiping out workspace first.
12:36:12 jenkins.util.io.CompositeIOException: Unable to delete '\gopath\src\github.com\docker\docker'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
```

tianon reviewed (approved):
LGTM

thaJeztah commented:
Boo! Why is it failing again?

```
14:40:56 W: Failed to fetch http://deb.debian.org/debian/dists/jessie/InRelease  
14:40:56 
14:40:56 W: Failed to fetch http://security.debian.org/debian-security/dists/jessie/updates/InRelease  
14:40:56 
14:40:56 W: Failed to fetch http://deb.debian.org/debian/dists/jessie/Release.gpg  Temporary failure resolving 'deb.debian.org'
14:40:56 
14:40:56 W: Failed to fetch http://security.debian.org/debian-security/dists/jessie/updates/Release.gpg  Temporary failure resolving 'security.debian.org'
14:40:56 
14:40:56 W: Some index files failed to download. They have been ignored, or old ones used instead.
```

thaJeztah committed:
WIP Move docker-py tests first again

See if networking works if we run it first

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Moved docker-py tests first again to see if the networking issues are because integration tests don't cleanup properly?

thaJeztah commented:
"YAY" so it passes if we run `docker-py` first

```
18:04:32 INFO: Building docker-sdk-python3:4.0.2...
18:14:10 sha256:c32514b539959ddb15e05f7c57552d0ca8d49ba25ac779f89663415c9b90a4c2
18:14:10 INFO: Starting docker-py tests...
18:14:33 ============================= test session starts ==============================
...
18:22:31 

18:22:31 = 367 passed, 10 skipped, 3 deselected, 3 xfailed, 2 xpassed in 477.76 seconds =
```

and it fails if we run it after `test-integration-flaky`, `test-integration`, and `cross`

![flippy](https://user-images.githubusercontent.com/1804568/61320332-80c3d300-a809-11e9-8e38-af83f88610e6.gif)

andrewhsu reviewed (approved):
SGTM w.r.t. getting more control over flaky tests

thaJeztah commented:
I'll update the last commit's title

seemethere commented:
Does it make sense if we even removed the `docker-py` tests from this repository completely?

thaJeztah commented:
@seemethere they have caught API regressions / changes at several occasions, so they can be useful. They should run separately though (I guess can do that once we have the Jenkinsfile up and running)

thaJeztah commented:
And, possibly, we should just have an image that we pull instead of building here 🤷‍♂
